### PR TITLE
[DDP] Add a correctness test case

### PR DESCRIPTION
**Pull request by alanwaketan**
Summary:
This pull request adds a correctness test case that compare the parameters after each training step between a DDP + TPU setup and a single CPU setup. If the inputs and labels are sharded correctly, the stepped parameters should match each.

Test Plan:
PJRT_DEVICE=TPU python test/pjrt/test_ddp.py TestPjRtDistributedDataParallel.test_ddp_correctness

### CODE REVIEW

File: test/pjrt/test_ddp.py [hunk 1]
```diff
@@ -16,7 +19,23 @@ def _init_xla_backend(init_file: str):
       init_method=f"file://{init_file}",
       rank=rank,
       world_size=world_size)
+  return rank, world_size
 
+def _assert_all_close(parameters_a, parameters_b):
```
Comment (will-cromar) [inline test/pjrt/test_ddp.py:25]: Will [`numpy.testing.assert_allclose`](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html#numpy.testing.assert_allclose) or [`absltest`'s `assertSequenceAlmostEqual`](https://github.com/abseil/abseil-py/blob/a0ae31683e6cf3667886c500327f292c893a1740/absl/testing/absltest.py#L1000-L1021) work here instead?

File: test/pjrt/test_ddp.py [hunk 2]
```diff
@@ -31,6 +50,49 @@ def _ddp_init(init_file: str):
...
+      ddp_labels = copy.deepcopy(cpu_labels[offset : offset + local_batch_size]).to(device)
+      ddp_loss = _train_step(ddp_model, ddp_inputs, ddp_labels, ddp_optimizer, loss_fn)
+      with torch.no_grad():
+        ddp_loss = ddp_loss / world_size
+        dist.all_reduce(ddp_loss)
+
+      # TODO(@alanwaketan): Investigate why the atol here is this low.
+      assert torch.allclose(cpu_loss, ddp_loss, atol=1e-02)
+      _assert_all_close(cpu_model.parameters(), ddp_model.parameters())
+      print(f"iteration {step}: cpu_loss = {cpu_loss}, ddp_loss = {ddp_loss}, cpu_model.parameters() ~= ddp_model.parameters()")
```
Comment (will-cromar) [inline test/pjrt/test_ddp.py]: Will this print 100 times during the unit test? Can you instead print only the final value or guard this behind some verbosity flag?

**Review by JackCaoG — COMMENTED** [review]
On CI, this test is being run as 1 CPU, I felt like that won't do real dpp check? https://github.com/pytorch/xla/blob/b7f54864a0933d6401a023e12ec6ccac1982e2d9/test/run_tests.sh#L105

@will-cromar I think PJRT multi CPU does not work?

**Review by will-cromar — APPROVED** [review]
Otherwise LGTM. Thanks!

**Comment by will-cromar** [comment]
The CI only runs one CPU with PjRt until we work out how to use PjRt CPU with multiprocessing. The manual test on TPU is more representative of a real use case, but I think this is still useful to run in the CI as a sanity check until then.

**Review by JackCaoG — APPROVED** [review]

File: test/pjrt/test_ddp.py (hunk 1 above)
Comment (alanwaketan) [reply test/pjrt/test_ddp.py:25]: `nn.module.parameters()` is a generator. Therefore, normal array check helpers won't work. https://pytorch.org/docs/stable/_modules/torch/nn/modules/module.html#Module.parameters.

File: test/pjrt/test_ddp.py (hunk 2 above)
Comment (alanwaketan) [reply test/pjrt/test_ddp.py]: Good point. I was too lazy. Let me add that.

**Comment by alanwaketan** [comment]
Thanks Jack and Will for approving this pull request.
